### ISSUE TITLE
Detecting communications layer down

### ISSUE THREAD

**Issue by LADSoft** [chkr1011/MQTTnet#1033]
### Describe your question
on TCP we are accustomed to being able to determine if the communications layer goes down, for example the other side closes the socket and sends RST next time you try to send it something.    MQTTNet doesn't seem to deal with this though, we don't seem to get a 'disconnect' event until the keep-alive fails and by then message send failures meant we had to notify the user there is a problem (which could have been smoothed over with the retry mechanism if we could have detected it soon enough and rebuilt the connection).    Can you verify this is indeed how MQTTNet works?    And if so is there any workaround other than set the keepalive to a very low value?

### Which project is your question related to?
- Client
- Server

**Comment by chkr1011** [comment]
My understanding is that the RST message will result in a graceful close and MQTTnet should directly throw an exception and stop the receiver task. 

Which platform are you using? UWP, NETSTANDARD, NET Framework 4.5.2 etc.?

**Comment by LADSoft** [comment]
Net framework 4.7.    AFAICT I'm not getting a 'disconnect' event from MQTTNet...

**Comment by LADSoft** [comment]
I was originally polling _client.IsConnected(), but read somewhere that isn't recommended.    Should I have been using the disconnect event from the beginning?   Does it really make a difference?

**Comment by chkr1011** [comment]
The property _IsConnected_ and the _Disconnected_ event are basically the same thing. The event is fired when the connection is closed and thus the _IsConnected_ turns into _False_. But unfortunately this seems to be not detected in your case.

What makes we wonder is that there is a Unit Tests which basically has the same problem. When I send data with the Low Level MQTT client the socket throws no exception at all even if the server is already down. From my understanding we should get an exception because the client is never getting an TCP ACK for the frames which are sent....

**Comment by LADSoft** [comment]
I now read some stuff on stackoverflow and the claim is you can't do it without pings...   which I would believe about the TCP ACK since the stack has to have a very long timeout for TCP ACK.   In case there is a real-world communications lag.    Still skeptical about the RST though, seems like THAT should be passable because it is an obvious break in communications, and I know in C/C++ code it is indeed passed out of the TCP stack...

We're specifically concerned with the situation where TCP connections just randomly go south at odd intervals but you can remake them immediately with no problem.   The major issue being we have to annunciate communication breaks but annunciating random disconnects that we can easily resolve in the background really annoys our users.   These breaks do happen frequently enough in the real world to be a bother...  (bear in mind our application is designed to be connected over TCP 24/7 for potentially years on end).   Ideally we would do the detection/remaking the connection quickly in the background.  In C/C++ we have done just by observing the results of a  send() but it is proving elusive in this case...

**Comment by LADSoft** [comment]
ok so i  grabbed a look at the MQTT code.   you have code similar to this in a couple of places:

```
                if (socketException.SocketErrorCode == SocketError.ConnectionAborted ||
                    socketException.SocketErrorCode == SocketError.OperationAborted)
                {
                    throw new OperationCanceledException();
                }

```

I'm assuming the throw out of that if block is what should cause the disconnect event to fire?   Possibly, that should check for _SocketError.ConnectionReset_ as well...   maybe some other stuff. I'm not really that familiar with WINSOCK errors...

**Comment by chkr1011** [comment]
This if is only there to throw the proper type of exception.
All other socket error codes are covered in the throw statement at the end of the catch block. So also other socket exceptions are stopping the client.

Could you please try with the RC version of 3.0.14? This version contains some changes in connection management.

**Comment by LADSoft** [comment]
I don't know if I can; we generally use NUGET for deployment.   I'll see if I can figure something out for temporary testing but it will be after thanksgiving is over before I can get to it.

**Comment by chkr1011** [comment]
@LADSoft There is also a nuget available of 3.0.14-RC1

**Comment by LADSoft** [comment]
ok thanks, will get to it as soon as I get back from the holidays :smile:

**Comment by LADSoft** [comment]
i got the nuget the other day and installed it; unfortunately didn't address the issue as there was a failure last night in the early morning hours...

**Comment by deandob** [comment]
I'm experiencing similar problems, I have a long term MQTT/Sockets connection to a remote broker that will randomly disconnect, and I'm not getting any disconnect events happening so we don't get the chance to reconnect in the background, basically making our application look like its offline.
LADSoft - did you find a resolution to this problem?

**Comment by LADSoft** [comment]
no i didn't...   we found a bug in our sources but I'm not really all that hopeful that it will correct this problem.    We are letting it run just in case....   I was thinking about waiting until we get VS2019 (we are stuck on VS2017 for the moment) then I can debug MQTTnet myself but in the meantime I have nothing...

**Comment by LADSoft** [comment]
@deandob we aren't able to duplicate at this time; going to leave it running over the holidays.   FWIW but bear in mind I don't know how relevant this is, one thing we changed on our end that seemed like it might have helped was to remake the connection if the 'publish' method threw exceptions that we weren't expecting...    but we also have a heartbeat that forces a publish at regular intervals.   Obviously publish isn't going to throw an exception if you don't call it...

**Comment by caglargul52** [comment]
I am having the same problem. Did you find a solution?

https://github.com/chkr1011/MQTTnet/issues/1054

**Comment by LADSoft** [comment]
I dunno.   I've been on holiday;   I left it running when I took off.   It had been running a few days without incident but not sure whether that means anything.    When I get back monday I will check to see where we stand....

**Comment by LADSoft** [comment]
ok so except for a glitch at the very beginning of the holiday which may have been due to other causes, this ran ok for about ten days for us.     It was failing more frequenly than that before.

I think but am not sure that the major contributor to it working now rather than before was that we have code to send pings every so often.   But there was a bug in that code wherein if an exception occurred on the send we didn't handle it.   

Disconnects were happening often enough that we were able to prove we could get such an exception if we sent something after there was a disconnect.    I think once we started handling the exception on send as a disconnect instead of waiting solely for the 'disconnect' event we were able to massage things a little better.    But not 100% sure on this as other seemingly unrelated things were fixed around the same time.
